<?xml version="1.0" encoding="utf-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="Docutils 0.11: http://docutils.sourceforge.net/" />
<title>Recipe 008: Do diversity analysis using IGS</title>
<style type="text/css">

/*
:Author: David Goodger (goodger@python.org)
:Id: $Id: html4css1.css 7614 2013-02-21 15:55:51Z milde $
:Copyright: This stylesheet has been placed in the public domain.

Default cascading style sheet for the HTML output of Docutils.

See http://docutils.sf.net/docs/howto/html-stylesheets.html for how to
customize this style sheet.
*/

/* used to remove borders from tables and images */
.borderless, table.borderless td, table.borderless th {
  border: 0 }

table.borderless td, table.borderless th {
  /* Override padding for "table.docutils td" with "! important".
     The right padding separates the table cells. */
  padding: 0 0.5em 0 0 ! important }

.first {
  /* Override more specific margin styles with "! important". */
  margin-top: 0 ! important }

.last, .with-subtitle {
  margin-bottom: 0 ! important }

.hidden {
  display: none }

a.toc-backref {
  text-decoration: none ;
  color: black }

blockquote.epigraph {
  margin: 2em 5em ; }

dl.docutils dd {
  margin-bottom: 0.5em }

object[type="image/svg+xml"], object[type="application/x-shockwave-flash"] {
  overflow: hidden;
}

/* Uncomment (and remove this text!) to get bold-faced definition list terms
dl.docutils dt {
  font-weight: bold }
*/

div.abstract {
  margin: 2em 5em }

div.abstract p.topic-title {
  font-weight: bold ;
  text-align: center }

div.admonition, div.attention, div.caution, div.danger, div.error,
div.hint, div.important, div.note, div.tip, div.warning {
  margin: 2em ;
  border: medium outset ;
  padding: 1em }

div.admonition p.admonition-title, div.hint p.admonition-title,
div.important p.admonition-title, div.note p.admonition-title,
div.tip p.admonition-title {
  font-weight: bold ;
  font-family: sans-serif }

div.attention p.admonition-title, div.caution p.admonition-title,
div.danger p.admonition-title, div.error p.admonition-title,
div.warning p.admonition-title, .code .error {
  color: red ;
  font-weight: bold ;
  font-family: sans-serif }

/* Uncomment (and remove this text!) to get reduced vertical space in
   compound paragraphs.
div.compound .compound-first, div.compound .compound-middle {
  margin-bottom: 0.5em }

div.compound .compound-last, div.compound .compound-middle {
  margin-top: 0.5em }
*/

div.dedication {
  margin: 2em 5em ;
  text-align: center ;
  font-style: italic }

div.dedication p.topic-title {
  font-weight: bold ;
  font-style: normal }

div.figure {
  margin-left: 2em ;
  margin-right: 2em }

div.footer, div.header {
  clear: both;
  font-size: smaller }

div.line-block {
  display: block ;
  margin-top: 1em ;
  margin-bottom: 1em }

div.line-block div.line-block {
  margin-top: 0 ;
  margin-bottom: 0 ;
  margin-left: 1.5em }

div.sidebar {
  margin: 0 0 0.5em 1em ;
  border: medium outset ;
  padding: 1em ;
  background-color: #ffffee ;
  width: 40% ;
  float: right ;
  clear: right }

div.sidebar p.rubric {
  font-family: sans-serif ;
  font-size: medium }

div.system-messages {
  margin: 5em }

div.system-messages h1 {
  color: red }

div.system-message {
  border: medium outset ;
  padding: 1em }

div.system-message p.system-message-title {
  color: red ;
  font-weight: bold }

div.topic {
  margin: 2em }

h1.section-subtitle, h2.section-subtitle, h3.section-subtitle,
h4.section-subtitle, h5.section-subtitle, h6.section-subtitle {
  margin-top: 0.4em }

h1.title {
  text-align: center }

h2.subtitle {
  text-align: center }

hr.docutils {
  width: 75% }

img.align-left, .figure.align-left, object.align-left {
  clear: left ;
  float: left ;
  margin-right: 1em }

img.align-right, .figure.align-right, object.align-right {
  clear: right ;
  float: right ;
  margin-left: 1em }

img.align-center, .figure.align-center, object.align-center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}

.align-left {
  text-align: left }

.align-center {
  clear: both ;
  text-align: center }

.align-right {
  text-align: right }

/* reset inner alignment in figures */
div.align-right {
  text-align: inherit }

/* div.align-center * { */
/*   text-align: left } */

ol.simple, ul.simple {
  margin-bottom: 1em }

ol.arabic {
  list-style: decimal }

ol.loweralpha {
  list-style: lower-alpha }

ol.upperalpha {
  list-style: upper-alpha }

ol.lowerroman {
  list-style: lower-roman }

ol.upperroman {
  list-style: upper-roman }

p.attribution {
  text-align: right ;
  margin-left: 50% }

p.caption {
  font-style: italic }

p.credits {
  font-style: italic ;
  font-size: smaller }

p.label {
  white-space: nowrap }

p.rubric {
  font-weight: bold ;
  font-size: larger ;
  color: maroon ;
  text-align: center }

p.sidebar-title {
  font-family: sans-serif ;
  font-weight: bold ;
  font-size: larger }

p.sidebar-subtitle {
  font-family: sans-serif ;
  font-weight: bold }

p.topic-title {
  font-weight: bold }

pre.address {
  margin-bottom: 0 ;
  margin-top: 0 ;
  font: inherit }

pre.literal-block, pre.doctest-block, pre.math, pre.code {
  margin-left: 2em ;
  margin-right: 2em }

pre.code .ln { color: grey; } /* line numbers */
pre.code, code { background-color: #eeeeee }
pre.code .comment, code .comment { color: #5C6576 }
pre.code .keyword, code .keyword { color: #3B0D06; font-weight: bold }
pre.code .literal.string, code .literal.string { color: #0C5404 }
pre.code .name.builtin, code .name.builtin { color: #352B84 }
pre.code .deleted, code .deleted { background-color: #DEB0A1}
pre.code .inserted, code .inserted { background-color: #A3D289}

span.classifier {
  font-family: sans-serif ;
  font-style: oblique }

span.classifier-delimiter {
  font-family: sans-serif ;
  font-weight: bold }

span.interpreted {
  font-family: sans-serif }

span.option {
  white-space: nowrap }

span.pre {
  white-space: pre }

span.problematic {
  color: red }

span.section-subtitle {
  /* font-size relative to parent (h1..h6 element) */
  font-size: 80% }

table.citation {
  border-left: solid 1px gray;
  margin-left: 1px }

table.docinfo {
  margin: 2em 4em }

table.docutils {
  margin-top: 0.5em ;
  margin-bottom: 0.5em }

table.footnote {
  border-left: solid 1px black;
  margin-left: 1px }

table.docutils td, table.docutils th,
table.docinfo td, table.docinfo th {
  padding-left: 0.5em ;
  padding-right: 0.5em ;
  vertical-align: top }

table.docutils th.field-name, table.docinfo th.docinfo-name {
  font-weight: bold ;
  text-align: left ;
  white-space: nowrap ;
  padding-left: 0 }

/* "booktabs" style (no vertical lines) */
table.docutils.booktabs {
  border: 0px;
  border-top: 2px solid;
  border-bottom: 2px solid;
  border-collapse: collapse;
}
table.docutils.booktabs * {
  border: 0px;
}
table.docutils.booktabs th {
  border-bottom: thin solid;
  text-align: left;
}

h1 tt.docutils, h2 tt.docutils, h3 tt.docutils,
h4 tt.docutils, h5 tt.docutils, h6 tt.docutils {
  font-size: 100% }

ul.auto-toc {
  list-style-type: none }

</style>
</head>
<body>
<div class="document" id="recipe-008-do-diversity-analysis-using-igs">
<h1 class="title">Recipe 008: Do diversity analysis using IGS</h1>

<p>This recipe shows the procedure to do diversity analysis using the concept of
IGS, for whole shotgun metagenomic sequencing data.</p>
<p>Basically instead of using OTUs, IGSs(informative genomic segment) are used as
the cornerstone to evaluate diversity. The recipe mainly shows the pipeline
to generate the IGSs-by-samples matrix from raw sequencing reads. After
we get the IGSs-
by-samples matrix, we use QIIME to do the alpha/beta diversity analysis.</p>
<p>For this recipe, firstly we create 6 synthetic samples (Sample 1-6)
based on 9 synthetic 10K
genomes (genome A-I), with different composition of species and diversity.</p>
<p>The sequencing depth of all the synthetic data sets is 10X. For simplicity,
there is no sequencing errors introduced in the synthetic reads data sets.</p>
<p>The final species abundance for each synthetic sample is as below:</p>
<pre class="literal-block">
sample1: genomeA -&gt; 30, genomeB -&gt; 10

sample2: genomeA -&gt; 20, genomeB -&gt; 10, genomeC -&gt; 10

sample3: genomeA -&gt; 10, genomeB -&gt; 10, genomeC -&gt; 10, genomeD -&gt; 10

sample4: genomeA -&gt; 10, genomeB -&gt; 10, genomeC -&gt; 10, genomeE -&gt; 10

sample5: genomeA -&gt; 10, genomeF -&gt; 10, genomeG -&gt; 10, genomeH -&gt; 10

sample6: genomeI -&gt; 10, genomeF -&gt; 10, genomeG -&gt; 10, genomeH -&gt; 10
</pre>
<p>The synthetic sequencing data sets for the 6 samples are sample_1.fa, sample_2
.fa, sample_3.fa, etc.</p>
<!-- shell start -->
<!-- ::

. ~/dev/ipy7/bin/activate
set -e

# create 9 simulated genomes.
python ~/dev/nullgraph/make-random-genome.py -l 10000 -s 1  > genomeA.fa
python ~/dev/nullgraph/make-random-genome.py -l 10000 -s 2  > genomeB.fa
python ~/dev/nullgraph/make-random-genome.py -l 10000 -s 3  > genomeC.fa
python ~/dev/nullgraph/make-random-genome.py -l 10000 -s 4  > genomeD.fa
python ~/dev/nullgraph/make-random-genome.py -l 10000 -s 5  > genomeE.fa
python ~/dev/nullgraph/make-random-genome.py -l 10000 -s 6  > genomeF.fa
python ~/dev/nullgraph/make-random-genome.py -l 10000 -s 7  > genomeG.fa
python ~/dev/nullgraph/make-random-genome.py -l 10000 -s 8  > genomeH.fa
python ~/dev/nullgraph/make-random-genome.py -l 10000 -s 9  > genomeI.fa


# build a read set
python ~/dev/nullgraph/make-reads.py -C 30 -e 0.00 genomeA.fa > reads_A_30.fa
python ~/dev/nullgraph/make-reads.py -C 20 -e 0.00 genomeA.fa > reads_A_20.fa
python ~/dev/nullgraph/make-reads.py -C 10 -e 0.00 genomeA.fa > reads_A_10.fa
python ~/dev/nullgraph/make-reads.py -C 10 -e 0.00 genomeB.fa > reads_B_10.fa
python ~/dev/nullgraph/make-reads.py -C 10 -e 0.00 genomeC.fa > reads_C_10.fa
python ~/dev/nullgraph/make-reads.py -C 10 -e 0.00 genomeD.fa > reads_D_10.fa
python ~/dev/nullgraph/make-reads.py -C 10 -e 0.00 genomeE.fa > reads_E_10.fa
python ~/dev/nullgraph/make-reads.py -C 10 -e 0.00 genomeF.fa > reads_F_10.fa
python ~/dev/nullgraph/make-reads.py -C 10 -e 0.00 genomeG.fa > reads_G_10.fa
python ~/dev/nullgraph/make-reads.py -C 10 -e 0.00 genomeH.fa > reads_H_10.fa
python ~/dev/nullgraph/make-reads.py -C 10 -e 0.00 genomeI.fa > reads_I_10.fa

# build the synthetic data sets for samples
cat reads_A_30.fa reads_B_10.fa >sample_1.fa
cat reads_A_20.fa reads_B_10.fa reads_C_10.fa >sample_2.fa
cat reads_A_10.fa reads_B_10.fa reads_C_10.fa reads_D_10.fa >sample_3.fa
cat reads_A_10.fa reads_B_10.fa reads_C_10.fa reads_E_10.fa >sample_4.fa
cat reads_A_10.fa reads_F_10.fa reads_G_10.fa reads_H_10.fa >sample_5.fa
cat reads_I_10.fa reads_F_10.fa reads_G_10.fa reads_H_10.fa >sample_6.fa -->
<p>The first thing we need do is to generate the hash table files for each
reads fasta file:</p>
<pre class="literal-block">
for i in sample_*.fa; do load-into-counting.py -x 1e6 -k 20 $i.kh $i ;  done
</pre>
<p>Next we use a script <tt class="docutils literal">get_comb_multi.py</tt> to generate the coverage of each read
across samples. For each sample reads fasta file, there is an output file
containing the coverage information.</p>
<p>But before we run the script, we need to generate a <tt class="docutils literal">config.txt</tt> file, which
contains three lines, the first line with a list of hashtable files, the
second line with a list of fasta files, and the third line with the size
of memory available to use.</p>
<pre class="literal-block">
ls sample_*.kh | awk '{ ORS=&quot; &quot;; print; }'&gt;config.txt
printf &quot;\n&quot; &gt;&gt;config.txt

ls sample_*.fa | awk '{ ORS=&quot; &quot;; print; }' &gt;&gt;config.txt
printf &quot;\n&quot; &gt;&gt;config.txt
printf &quot;30000000&quot; &gt;&gt;config.txt
</pre>
<p>Let's run the script:</p>
<pre class="literal-block">
./python get_comb_multi.py config.txt
</pre>
<p>Let's check an output file:</p>
<pre class="literal-block">
$ more sample_1.fa.comb
read0f_genomeA.fa 24 18 10 10 10 0
read1r_genomeA.fa 20 15 7 7 7 0
read2r_genomeA.fa 19 12 7 7 7 0
read3f_genomeA.fa 21 14 8 8 8 0
read4r_genomeA.fa 19 14 6 6 6 0
read5r_genomeA.fa 26 23 11 11 11 0
read6r_genomeA.fa 19 15 8 8 8 0
</pre>
<p>Here we can see for each read in this sample (sample_1), the coverage numbers
of this read across the 6 samples are listed in different columns. The order
of the samples is in accordance with the order of the files listed in the first
line of &quot;config.txt&quot; file.</p>
<p>Next we run <tt class="docutils literal">count_spectrum_freq_multiple_files.py</tt> to get the count of reads
with each coverage spectrum across different samples.
A &quot;coverage spectrum&quot; is the coverage of a read
across different samples. A read with a coverage spectrum as &quot;1-0-0-0-2-1&quot;
has a coverage as &quot;1&quot; in sample1, is not covered in sample2,3,4, has a coverage
as &quot;2&quot; in sample5, has a coverage as &quot;1&quot; in sample 6 at last.</p>
<p>To run <tt class="docutils literal">count_spectrum_freq_multiple_files.py</tt>, a file with a list of the
&quot;.comb&quot; files should be generated beforehand:</p>
<pre class="literal-block">
ls *.comb &gt;comb.list
python count_spectrum_freq_multiple_files.py comb.list all_sample.spectrum
</pre>
<p>An except from output file:</p>
<pre class="literal-block">
0-0-0-0-10-10 0 0 0 0 372 372
0-0-0-0-11-11 0 0 0 0 249 249
0-0-0-0-12-12 0 0 0 0 252 252
0-0-0-0-13-13 0 0 0 0 216 216
0-0-0-0-14-14 0 0 0 0 66 66
0-0-0-0-15-15 0 0 0 0 9 9
0-0-0-0-16-16 0 0 0 0 36 36
</pre>
<div class="section" id="beta-diversity-analysis">
<h1>Beta-diversity analysis</h1>
<p>This shows there are 372 reads in sample 5 and 372 reads in sample 5 with
the coverage spectrum of &quot;0-0-0-0-10-10&quot;. In other way, 372 reads in sample 5
and sample 6 respectfully have a coverage as 10 in both samples. Here because
there is no sequencing error in data sets, so the number of reads in both
with same coverage spectrum is the same. (the ratio of 10:10)
If there is error, it may be different, but the ratio should be similar
to the ratio of the coverage in the spectrum. Divided 372
by 10, we can say 37 (we round 37.2 to 37) IGSs are shared by sample 5 and
sample 6. Actually, the calculation should be (372+372)/(10+10). But in this
case, no big difference. All the IGSs have a coverage as 10 in both samples.</p>
<p>Next we can list all the IGSs and corresponding coverage(abundance) across
samples to get the IGSs-by-samples matrix.</p>
<p>Here we need a MAP file to store the meta data of the data set,
which is also required by QIIME. Since this is synthetic data set, the MAP
file is pretty simple:</p>
<pre class="literal-block">
$ more all_sample_MAP.txt
#SampleID       BarcodeSequence LinkerPrimerSequence    Description
sample1 A       A       AAAB
sample2 A       A       AABC
sample3 A       A       ABCD
sample4 A       A       ABCE
sample5 A       A       AFGH
sample6 A       A       IFGH
</pre>
<pre class="literal-block">
python seperate_IGS.py all_sample.spectrum all_sample_MAP.txt
</pre>
<p>An except from output file:</p>
<pre class="literal-block">
202     0       0       0       0       13      13
203     0       0       0       0       13      13
204     0       0       0       0       13      13
205     0       0       0       0       13      13
206     0       0       0       0       13      13
207     0       0       0       0       13      13
208     0       0       0       0       13      13
209     0       0       0       0       14      14
210     0       0       0       0       14      14
211     0       0       0       0       14      14
212     0       0       0       0       14      14
213     0       0       0       0       16      16
214     0       0       0       0       16      16
215     0       0       0       0       3       3
216     0       0       0       0       3       3
217     0       0       0       0       3       3
218     0       0       0       0       3       3
219     0       0       0       0       4       4
</pre>
<p>This output file lists all the IGSs and the corresponding abundance across
the 6 samples in each line.</p>
<p>Next we can use QIIME to do the beta analysis. Before that, we need to
convert the IGSs-by-samples matrix into the BIOM format:</p>
<pre class="literal-block">
biom convert -i  all_sample.spectrum.IGS -o all_sample.spectrum.IGS.biom --table-type=&quot;OTU table&quot;
</pre>
<p>Also, we get some statistics from the BIOM file:</p>
<pre class="literal-block">
biom summarize-table -i all_sample.spectrum.IGS.biom -o all_sample.spectrum.IGS.biom.summary.txt
</pre>
<p>Output summary file contains the number of counts in different sample. We
will need the information for the analysis below:</p>
<pre class="literal-block">
Counts/sample detail:
 sample1: 1057.0
 sample2: 1956.0
 sample3: 2858.0
 sample4: 2858.0
 sample5: 2988.0
 sample6: 3886.0
</pre>
<p>With the BIOM file and MAP file, we can use QIIME pipeline to do beta diversity
analysis. See <a class="reference external" href="http://qiime.org/tutorials/tutorial.html">http://qiime.org/tutorials/tutorial.html</a> for details.</p>
<dl class="docutils">
<dt>Here we use beta diversity metrics as &quot;bray curtis&quot;, which is set in preference</dt>
<dd>file <tt class="docutils literal">p_file.txt</tt>.</dd>
</dl>
<div class="section" id="compute-beta-diversity-and-generate-beta-diversity-plots">
<h2>Compute Beta Diversity and Generate Beta Diversity Plots</h2>
<pre class="literal-block">
beta_diversity_through_plots.py -i all_sample.spectrum.IGS.biom -o bdiv_even1000/ -m all_sample_MAP.txt -e 1000 -p p_file.txt
</pre>
<p>We can open the <tt class="docutils literal">index.html</tt> file in folder <tt class="docutils literal">bdiv_even1000/bray_curtis_emperor_pcoa_plot/</tt>
to see the interactive PCA figure, like shown below.</p>
<img alt="PCA_3d.png" src="PCA_3d.png" style="width: 500px;" />
<p>We can also call a seperate script to draw 2D plots:</p>
<pre class="literal-block">
make_2d_plots.py -i bdiv_even1000/bray_curtis_pc.txt -m all_sample_MAP.txt -o beta_2d_plots/
</pre>
<img alt="PC1_vs_PC2_plot.png" src="PC1_vs_PC2_plot.png" style="width: 500px;" />
</div>
<div class="section" id="jackknifed-beta-diversity-and-hierarchical-clustering">
<h2>Jackknifed Beta Diversity and Hierarchical Clustering</h2>
<pre class="literal-block">
jackknifed_beta_diversity.py -i all_sample.spectrum.IGS.biom -o bdiv_jk1000 -e 1000 -m  all_sample_MAP.txt -p p_file.txt
make_bootstrapped_tree.py -m bdiv_jk1000/bray_curtis/upgma_cmp/master_tree.tre -s bdiv_jk1000/bray_curtis/upgma_cmp/jackknife_support.txt -o bdiv_jk1000/bray_curtis/upgma_cmp/jackknife_named_nodes.pdf
</pre>
<img alt="tree.png" src="tree.png" style="width: 500px;" />
</div>
</div>
<div class="section" id="alpha-diversity-analysis">
<h1>Alpha-diversity analysis</h1>
<p>For alpha-diversity, the procedure is similar to beta-diversity shown above.
The difference is that for higher accuracy, we treat each sample seperately and
get the abundance distribution of IGSs in each sample.</p>
<p>To do this, instead of using script <tt class="docutils literal">seperate_IGS.py</tt>, we use script
<tt class="docutils literal">seperate_IGS_for_alpha.py</tt> to list the IGSs out:</p>
<pre class="literal-block">
python seperate_IGS_for_alpha.py all_sample.spectrum all_sample_MAP.txt
</pre>
<p>Next, the same as beta-diversity procedure:</p>
<pre class="literal-block">
biom convert -i all_sample.spectrum.IGS.alpha -o all_sample.spectrum.IGS.alpha.biom --table-type=&quot;OTU table&quot;
biom summarize-table -i  all_sample.spectrum.IGS.alpha.biom -o all_sample.spectrum.IGS.alpha.biom.summary.txt
</pre>
<p>We'd like to calculate chao1 estimator, so we create a parameter file firstly:</p>
<pre class="literal-block">
echo &quot;alpha_diversity:metrics chao1,observed_species&quot; &gt; alpha_params.txt
</pre>
<p>Then run alpha diversity pipeline:</p>
<pre class="literal-block">
alpha_rarefaction.py -i all_sample.spectrum.IGS.alpha.biom -m all_sample_MAP.txt -o wf_arare/ -p alpha_params.txt -f
</pre>
<p>Observed IGSs:</p>
<img alt="observed_alpha.png" src="observed_alpha.png" style="width: 500px;" />
<p>Chao1 estimator:</p>
<img alt="chao_alpha.png" src="chao_alpha.png" style="width: 500px;" />
<div class="section" id="resources-and-links">
<h2>Resources and Links</h2>
<p><a class="reference external" href="https://github.com/ged-lab/khmer-recipes/tree/master/004-estimate-sequencing-saturation">This recipe</a>
is hosted in the khmer-recipes repository,
<a class="reference external" href="https://github.com/ged-lab/khmer-recipes/">https://github.com/ged-lab/khmer-recipes/</a>.</p>
<p>It requires the <a class="reference external" href="http://khmer.readthedocs.org">khmer software</a>.</p>
</div>
</div>
</div>
</body>
</html>
